File: apple/App/JoinLink.swift
import Foundation

/// Doc utilisateur, révisé P9 — code d'appairage encodé en lien, pour éviter la saisie manuelle.
/// Contenu minimal : le code d'appairage lui-même — `SupabaseTransport.resolveGame(code:)` résout
/// tout le reste (matchID, jeu, nombre de joueurs, nom de l'appareil) à partir de ce seul code, ce
/// lien ne fait que remplacer la frappe des 6 chiffres.
///
/// Schéma personnalisé (`quimene://`) plutôt qu'un lien universel `https://` : ce dernier
/// demanderait de posséder un nom de domaine et d'y héberger un fichier de vérification
/// (Associated Domains/App Links), une dépendance externe hors de portée pour l'instant. En
/// échange, l'ouverture depuis l'appareil photo système n'est garantie que sur iOS (Camera
/// propose « Ouvrir dans Qui Mène ? » pour un schéma personnalisé si l'app est installée) ; le
/// scanner intégré (`QRScannerView`) reste le chemin fiable sur toutes les plateformes.
enum JoinLink {
  private static let scheme = "quimene"
  private static let host = "join"

  struct Payload: Equatable {
    let pairingCode: String
  }

  static func url(pairingCode: String) -> URL? {
    var components = URLComponents()
    components.scheme = scheme
    components.host = host
    components.queryItems = [
      URLQueryItem(name: "code", value: pairingCode)
    ]
    return components.url
  }

  static func parse(_ url: URL) -> Payload? {
    guard url.scheme == scheme, url.host == host,
      let components = URLComponents(url: url, resolvingAgainstBaseURL: false),
      let items = components.queryItems,
      let code = items.first(where: { $0.name == "code" })?.value
    else { return nil }
    return Payload(pairingCode: code)
  }
}

File: android/app/src/main/kotlin/com/quimene/app/livesync/JoinLink.kt
package com.quimene.app.livesync

import android.net.Uri

/**
 * Miroir de `JoinLink.swift` — le lien encodé dans le QR de partage (`QrCodeView` côté hôte) et
 * décodé au scan (`QrScannerView` côté pair). Contenu minimal : le code d'appairage lui-même,
 * `SupabaseSessionBackend.resolve(code)`
 * résout tout le reste — ce lien ne fait que remplacer la frappe des 6 chiffres. Même schéma
 * personnalisé `quimene://join?code=XXXXXX` qu'Apple, pas un lien universel `https://` (même
 * raison : pas de nom de domaine à posséder/vérifier) — un code généré par l'un ou l'autre app
 * reste scannable par l'autre plateforme.
 */
object JoinLink {
    private const val SCHEME = "quimene"
    private const val HOST = "join"

    fun url(pairingCode: String): String = "$SCHEME://$HOST?code=$pairingCode"

    fun parse(raw: String): String? {
        val uri = runCatching { Uri.parse(raw) }.getOrNull() ?: return null
        if (uri.scheme != SCHEME || uri.host != HOST) return null
        return uri.getQueryParameter("code")
    }
}
